### PR TITLE
feat: Array of parameters deserialization

### PR DESCRIPTION
**Pull request by EwenQuim**

### CODE REVIEW

**Comment by EwenQuim** [comment]
@dylanhitt 🎁

**Comment by dylanhitt** [comment]
Will look at this and the other tomorrow.

File: ctx.go [hunk 1]
```diff
@@ -376,54 +411,47 @@ func (c *netHttpContext[B, P]) Params() (P, error) {
...
-			}
-		} else {
-			tag = field.Tag.Get("header")
-			if tag != "" {
-				paramValue = c.Header(tag)
-				if paramValue == "" {
+		// Process query parameters
+		if tag := field.Tag.Get("query"); tag != "" {
+			// Handle slice/array types
+			if field.Type.Kind() == reflect.Slice || field.Type.Kind() == reflect.Array {
```
Comment (dylanhitt) [inline ctx.go:419]: Nit pick: I tend to like switch cases when I see this

File: ctx.go [hunk 2]
```diff
@@ -376,54 +411,47 @@ func (c *netHttpContext[B, P]) Params() (P, error) {
...
+				if err != nil {
+					return *p, err
+				}
 			}
-			fieldValue.SetUint(uintValue)
-		case reflect.Float32, reflect.Float64:
-			floatValue, err := strconv.ParseFloat(paramValue, 64)
-			if err != nil {
-				return *p, fmt.Errorf("cannot convert %s to float64: %w", paramValue, err)
+		} else if tag := field.Tag.Get("header"); tag != "" {
```
Comment (dylanhitt) [inline ctx.go:445]: Same here. But feel free disregard it's just a style thing. I'm find myself in the perpetual war against else statement

**Review by dylanhitt — APPROVED** [review]

File: ctx.go [hunk 3]
```diff
@@ -365,6 +365,41 @@ func (c *netHttpContext[B, P]) Body() (B, error) {
...
+		}
+		value.SetBool(boolValue)
+	default:
+		return fmt.Errorf("unsupported type %s", kind)
+	}
+	return nil
+}
+
 func (c *netHttpContext[B, P]) Params() (P, error) {
 	p := new(P)
```
Comment (ccoVeille) [inline ctx.go:404]: Nit: This is old code, but why creating a pointer, when method return the non-pointer version

File: ctx.go [hunk 4]
```diff
@@ -376,54 +411,47 @@ func (c *netHttpContext[B, P]) Params() (P, error) {
...
-			boolValue, err := strconv.ParseBool(paramValue)
+			err := setParamValue(fieldValue, paramValue, field.Type.Kind())
 			if err != nil {
-				return *p, fmt.Errorf("cannot convert %s to bool: %w", paramValue, err)
+				return *p, err
 			}
-			fieldValue.SetBool(boolValue)
-		default:
-			return *p, fmt.Errorf("unsupported type %s", field.Type.Kind())
 		}
```
Comment (ccoVeille) [inline ctx.go:455]: Nit: I would move all this if into its own method.

Or at least, I would return early when there is no tag.

The code is pretty complicated with a lot of branching

File: ctx.go (hunk 1 above)
Comment (ccoVeille) [reply ctx.go:419]: I approve this message 😁

File: ctx.go (hunk 2 above)
Comment (ccoVeille) [reply ctx.go:445]: I approve this message 😁

File: ctx_test.go [hunk 5]
```diff
@@ -690,4 +690,147 @@ func TestNetHttpContext_Params(t *testing.T) {
...
+			Int64s  []int64  `query:"int64s"`
+			Uints   []uint   `query:"uints"`
+			Uint8s  []uint8  `query:"uint8s"`
+			Uint16s []uint16 `query:"uint16s"`
+			Uint32s []uint32 `query:"uint32s"`
+			Uint64s []uint64 `query:"uint64s"`
+		}
+
+		url := "http://example.com/foo?" +
+			"int8s=1&int8s=2&" +
```
Comment (ccoVeille) [inline ctx_test.go:736]: Please also tests how the code reacts to overflow, like settings 1024 into an int8 that cannot go further than 128

File: ctx.go [hunk 6]
```diff
@@ -365,6 +365,41 @@ func (c *netHttpContext[B, P]) Body() (B, error) {
 	return body, err
 }
 
+// setParamValue sets a value to a reflect.Value based on its kind
+func setParamValue(value reflect.Value, paramValue string, kind reflect.Kind) error {
+	switch kind {
+	case reflect.String:
+		value.SetString(paramValue)
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		intValue, err := strconv.Atoi(paramValue)
```
Comment (ccoVeille) [inline ctx.go:374]: Please use  strconv.ParseInt

You also need to play with the base here

```suggestion
		intValue, err := strconv.ParseInt(paramValue, 10, bitSize(kind))
```

See https://github.com/go-fuego/fuego/pull/501#discussion_r2057705693

File: ctx.go [hunk 7]
```diff
@@ -365,6 +365,41 @@ func (c *netHttpContext[B, P]) Body() (B, error) {
...
+		if err != nil {
+			return fmt.Errorf("cannot convert %s to int: %w", paramValue, err)
+		}
+		value.SetInt(int64(intValue))
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		uintValue, err := strconv.ParseUint(paramValue, 10, 64)
+		if err != nil {
+			return fmt.Errorf("cannot convert %s to uint: %w", paramValue, err)
+		}
+		value.SetUint(uintValue)
```
Comment (ccoVeille) [inline ctx.go:384]: Here you are facing a problem with overflow

You should do this.

```suggestion
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		uintValue, err := strconv.ParseUint(paramValue, 10, bitSize(kind))
		if err != nil {
			return fmt.Errorf("cannot convert %s to uint: %w", paramValue, err)
		}
		value.SetUint(uintValue)
```

With a method that returns you the bit size

```go
func bitSize(kind) {
switch kind {
   case reflect.Uint8, reflect.Int8:
     return 8
   case reflect.Uint16, reflect.Int16:
     return 16
   case reflect.Uint32, reflect.Int32, reflect.Float32:
     return 32
   case reflect.Uint, reflect.Int:
     return strconv.IntSize()
}
return 64
}
```

I wrote my own lib about this issue

https://github.com/ccoVeille/go-safecast

I don't expect you to use it, it's just the fact it's not easy to deal with this.

File: ctx.go [hunk 8]
```diff
@@ -365,6 +365,41 @@ func (c *netHttpContext[B, P]) Body() (B, error) {
...
+	case reflect.Float32, reflect.Float64:
+		floatValue, err := strconv.ParseFloat(paramValue, 64)
```
Comment (ccoVeille) [inline ctx.go:386]: Same here, you have to check and use bitSize


```suggestion
		floatValue, err := strconv.ParseFloat(paramValue, bitSize(kind))
```

**Review by ccoVeille — CHANGES_REQUESTED** [review]

File: ctx.go (hunk 1 above)
Comment (EwenQuim) [reply ctx.go]: Fixed here

File: ctx.go (hunk 2 above)
Comment (EwenQuim) [reply ctx.go:460]: I don't like switch for non-scalar comparisons 😕

File: ctx.go [hunk 9]
```diff
@@ -365,6 +365,41 @@ func (c *netHttpContext[B, P]) Body() (B, error) {
...
+	case reflect.Bool:
+		boolValue, err := strconv.ParseBool(paramValue)
+		if err != nil {
+			return fmt.Errorf("cannot convert %s to bool: %w", paramValue, err)
+		}
+		value.SetBool(boolValue)
+	default:
+		return fmt.Errorf("unsupported type %s", kind)
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (EwenQuim) [reply ctx.go:414]: No it isn't, there are no `return` clauses in the switch cases

[the remark this answers is not in the document]
Comment (ccoVeille) [reply ctx.go:414]: Oh indeed !

File: ctx.go [hunk 10]
```diff
@@ -365,6 +365,55 @@ func (c *netHttpContext[B, P]) Body() (B, error) {
...
+func setParamValue(value reflect.Value, paramValue string, kind reflect.Kind) error {
+	switch kind {
+	case reflect.String:
+		value.SetString(paramValue)
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		intValue, err := strconv.ParseInt(paramValue, 10, bitSize(kind))
+		if err != nil {
+			return fmt.Errorf("cannot convert %s to int: %w", paramValue, err)
+		}
+		value.SetInt(intValue)
```
Comment (ccoVeille) [inline ctx.go:392]: Just found this

https://pkg.go.dev/reflect#Value.OverflowInt

I would like to suggest this

```suggestion
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		intValue, err := strconv.ParseInt(paramValue, 10, 64)
		if err != nil {
			return fmt.Errorf("cannot convert %s to %s: %w", paramValue, kind, err)
		}
		if value.OverflowInt(intValue) {
			return fmt.Errorf("%s overflows %s", paramValue, kind)
		}
		value.SetInt(intValue)
```

I didn't test

File: ctx.go [hunk 11]
```diff
@@ -365,6 +365,55 @@ func (c *netHttpContext[B, P]) Body() (B, error) {
...
+		if err != nil {
+			return fmt.Errorf("cannot convert %s to uint: %w", paramValue, err)
+		}
+		value.SetUint(uintValue)
+	case reflect.Float32, reflect.Float64:
+		floatValue, err := strconv.ParseFloat(paramValue, bitSize(kind))
+		if err != nil {
+			return fmt.Errorf("cannot convert %s to float64: %w", paramValue, err)
+		}
+		value.SetFloat(floatValue)
```
[the remark this answers is not in the document]
Comment (EwenQuim) [reply ctx.go:404]: OK no problem ;)

File: ctx.go (hunk 10 above)
Comment (EwenQuim) [reply ctx.go:392]: Interesting!
